tests/test_reweighting.py

```diff
@@ -0,0 +1,407 @@
+    delta_f(params) and grad(delta_f)(params) are consistent with analytical_delta_f function
+
+    Parameters
+    ----------
+    estimate_delta_f: Callable[[params], float]
+        estimated free energy difference as a
+        differentiable function of params
+    analytical_delta_f: Callable[[params], float]
+        exact free energy difference as a
+        differentiable function of params
+    ref_params: [2,] array
+        generate random parameters near ref_params
+    n_random_trials: int
+    atol: float
+        absolute tolerance
+    """
+
+    def sample_random_params():
+        mean = ref_params[0] + onp.random.rand()
+        log_sigma = ref_params[1] - onp.random.rand()
+        return np.array([mean, log_sigma])
+
+    f_hat, g_hat = value_and_grad(estimate_delta_f)(ref_params)
+    f_ref, g_ref = value_and_grad(analytical_delta_f)(ref_params)
+
+    onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
+    onp.testing.assert_allclose(g_hat, g_ref, atol=atol)
+
+    for _ in range(n_random_trials):
+        trial_params = sample_random_params()
+        f_hat, g_hat = value_and_grad(estimate_delta_f)(trial_params)
+        f_ref, g_ref = value_and_grad(analytical_delta_f)(trial_params)
+
+        onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
+        onp.testing.assert_allclose(g_hat, g_ref, atol=atol)
+
+
+def test_endpoint_reweighting_1d():
+    """assert that endpoint reweighting estimator for delta_f(params), grad(delta_f)(params) is accurate
+    on tractable 1D system"""
+    onp.random.seed(2022)
+
+    u_fxn, _, sample, reduced_free_energy = make_gaussian_testsystem()
+
+    ref_params = np.zeros(2)
+    ref_delta_f = reduced_free_energy(1.0, ref_params) - reduced_free_energy(0.0, ref_params)
+
+    n_samples = int(1e6)
+
+    samples_0 = sample(0, ref_params, n_samples)
+    samples_1 = sample(1, ref_params, n_samples)
+
+    vec_u = vmap(u_fxn, in_axes=(0, None, None))
+    vec_u_0_fxn = lambda xs, params: vec_u(xs, 0, params)
+    vec_u_1_fxn = lambda xs, params: vec_u(xs, 1, params)
+
+    estimate_delta_f = construct_endpoint_reweighting_estimator(
+        samples_0, samples_1, vec_u_0_fxn, vec_u_1_fxn, ref_params, ref_delta_f
+    )
+    analytical_delta_f = lambda params: reduced_free_energy(1.0, params) - reduced_free_energy(0.0, params)
+
+    assert_estimator_accurate(jit(estimate_delta_f), analytical_delta_f, ref_params, n_random_trials=10, atol=5e-3)
+
+
+def test_mixture_reweighting_1d():
+    """using a variety of free energy estimates (MBAR, TI, analytical) to obtain reference mixture weights,
+    assert that mixture reweighting estimator of delta_f(params), grad(delta_f)(params) is accurate
+    on tractable 1D system"""
+    onp.random.seed(2022)
+
+    u_fxn, normalized_u_fxn, sample, reduced_free_energy = make_gaussian_testsystem()
+
+    ref_params = np.ones(2)
+    n_windows = 10
+    lambdas = np.linspace(0, 1, n_windows)
+
+    n_samples_per_window = int(1e5)
```

Comment:
Clarification - compared to the `endpoint_reweighting_test` above, we now have 10 windows that have 10x fewer steps, so the total number of samples is the same. However, the `atol=1e-2` is actually higher than the above test! This is counter-intuitive to me as I would have thought that use of intermediate windows should improve the efficiency of the estimator.

Reply:
>so the total number of samples is the same

Off by 2x. (endpoint test: 2 * 1e6, mixture test: 10 * 1e5)

>However, the atol=1e-2 is actually higher than the above test! This is counter-intuitive to me as I would have thought that use of intermediate windows should improve the efficiency of the estimator.

Initially I don't think this is surprising.

In the `endpoint_reweighting` case, we have `1e6` samples in lambda=0 state, and `1e6` samples in lambda=1 state.
In mixture case, we have at least `1e5` samples directly from lambda=0 state, but maybe we can borrow some effective samples from neighboring states. The total number of samples from all states is `1e6`. (So, we would only reach the same number of effective samples if all the other states were exactly equal to lambda=0 state... If the other states are different, we'll have somewhere in between `1e5` and `1e6` effective samples at lambda=0 state.)

If I instead fix the number of samples _per window_, then mixture reweighting should outperform endpoint reweighting. (Since the number of effective samples for estimating expectations at lambda=0 or lambda=1 would exceed the number of direct samples collected at lambda=0 or lambda=1.)